File: packages/python/src/kova_tts/websocket.py
from __future__ import annotations

import json
from dataclasses import asdict
from types import TracebackType
from typing import Any

from .errors import KovaTTSConnectionError, KovaTTSProtocolError
from .types import (
    AudioChunk,
    ContextClosed,
    ContextConfig,
    ContextStarted,
    ErrorFrame,
    FlushCompleted,
    Timestamps,
    TTSTimestamps,
    WebSocketFrame,
)


def serialize_start_context(
    *,
    voice_id: str,
    model_id: str,
    context_id: str | None = None,
    temperature: float | None = None,
    timestamps: bool | None = None,
) -> dict[str, Any]:
    config: dict[str, Any] = {
        "voice_id": voice_id,
        "model_id": model_id,
    }
    if temperature is not None:
        config["temperature"] = temperature
    if timestamps is not None:
        config["timestamps"] = timestamps

    frame: dict[str, Any] = {"start_context": config}
    if context_id is not None:
        frame["context_id"] = context_id
    return frame


def serialize_send_text(text: str, *, context_id: str | None = None) -> dict[str, Any]:
    frame: dict[str, Any] = {"send_text": text}
    if context_id is not None:
        frame["context_id"] = context_id
    return frame


def serialize_flush(
    *, context_id: str | None = None, flush_id: str | None = None
) -> dict[str, Any]:
    frame: dict[str, Any] = {"flush": True}
    if context_id is not None:
        frame["context_id"] = context_id
    if flush_id is not None:
        frame["flush_id"] = flush_id
    return frame


def serialize_close_context(
    *, context_id: str | None = None, flush_id: str | None = None
) -> dict[str, Any]:
    frame: dict[str, Any] = {"close_context": True}
    if context_id is not None:
        frame["context_id"] = context_id
    if flush_id is not None:
        frame["flush_id"] = flush_id
    return frame


def parse_websocket_frame(value: object) -> WebSocketFrame:
    if not isinstance(value, dict):
        raise KovaTTSProtocolError("WebSocket frame must be an object")

    if isinstance(value.get("error"), str):
        return ErrorFrame(
            error=value["error"],
            context_id=value.get("context_id"),
            flush_id=value.get("flush_id"),
            chunk_id=value.get("chunk_id"),
        )

    context_started = value.get("context_started")
    if isinstance(context_started, dict):
        return ContextStarted(
            context_started=_parse_context_config(context_started),
            context_id=value.get("context_id"),
        )

    if isinstance(value.get("audio_chunk"), str):
        return AudioChunk(
            audio_chunk=value["audio_chunk"],
            context_id=value.get("context_id"),
            chunk_id=value.get("chunk_id"),
        )

    timestamps = value.get("timestamps")
    if isinstance(timestamps, dict):
        return Timestamps(
            timestamps=_parse_timestamps(timestamps),
            context_id=value.get("context_id"),
            chunk_id=value.get("chunk_id"),
        )

    if value.get("flush_completed") is True and isinstance(value.get("flush_id"), str):
        return FlushCompleted(
            flush_completed=True,
            flush_id=value["flush_id"],
            context_id=value.get("context_id"),
            chunk_id=value.get("chunk_id"),
        )

    if value.get("context_closed") is True:
        return ContextClosed(context_closed=True, context_id=value.get("context_id"))

    raise KovaTTSProtocolError(f"Unknown WebSocket frame shape: {value!r}")


class KovaTTSWebSocket:
    def __init__(self, url: str, api_key: str):
        self._url = url
        self._api_key = api_key
        self._socket = None

    async def __aenter__(self) -> "KovaTTSWebSocket":
        try:
            import websockets

            try:
                self._socket = await websockets.connect(
                    self._url,
                    additional_headers={"x-api-key": self._api_key},
                )
            except TypeError:
                self._socket = await websockets.connect(
                    self._url,
                    extra_headers={"x-api-key": self._api_key},
                )
        except Exception as exc:
            raise KovaTTSConnectionError(str(exc)) from exc
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        traceback: TracebackType | None,
    ) -> None:
        if self._socket is not None:
            await self._socket.close()

    def __aiter__(self) -> "KovaTTSWebSocket":
        return self

    async def __anext__(self) -> WebSocketFrame:
        if self._socket is None:
            raise KovaTTSConnectionError("WebSocket is not connected")
        try:
            raw = await self._socket.recv()
        except StopAsyncIteration:
            raise
        except Exception as exc:
            if exc.__class__.__name__ == "ConnectionClosedOK":
                raise StopAsyncIteration from exc
            raise KovaTTSConnectionError(str(exc)) from exc

        frame = parse_websocket_frame(json.loads(raw))
        if isinstance(frame, ErrorFrame):
            raise KovaTTSProtocolError(frame.error)
        return frame

    async def start_context(
        self,
        *,
        voice_id: str,
        model_id: str,
        context_id: str | None = None,
        temperature: float | None = None,
        timestamps: bool | None = None,
    ) -> None:
        await self._send(
            serialize_start_context(
                voice_id=voice_id,
                model_id=model_id,
                context_id=context_id,
                temperature=temperature,
                timestamps=timestamps,
            )
        )

    async def send_text(self, text: str, *, context_id: str | None = None) -> None:
        await self._send(serialize_send_text(text, context_id=context_id))

    async def flush(self, *, context_id: str | None = None, flush_id: str | None = None) -> None:
        await self._send(serialize_flush(context_id=context_id, flush_id=flush_id))

    async def close_context(
        self, *, context_id: str | None = None, flush_id: str | None = None
    ) -> None:
        await self._send(serialize_close_context(context_id=context_id, flush_id=flush_id))

    async def _send(self, frame: dict[str, Any]) -> None:
        if self._socket is None:
            raise KovaTTSConnectionError("WebSocket is not connected")
        await self._socket.send(json.dumps(frame))


def websocket_frame_to_dict(frame: WebSocketFrame) -> dict[str, Any]:
    return asdict(frame)


def _parse_context_config(value: dict[str, Any]) -> ContextConfig:
    if not isinstance(value.get("voice_id"), str) or not isinstance(value.get("model_id"), str):
        raise KovaTTSProtocolError("Invalid context config")
    return ContextConfig(
        voice_id=value["voice_id"],
        model_id=value["model_id"],
        temperature=value.get("temperature"),
        timestamps=value.get("timestamps"),
    )


def _parse_timestamps(value: dict[str, Any]) -> TTSTimestamps:
    words = value.get("words")
    start_seconds = value.get("start_seconds")
    end_seconds = value.get("end_seconds")
    if not (isinstance(words, list) and isinstance(start_seconds, list) and isinstance(end_seconds, list)):
        raise KovaTTSProtocolError("Invalid timestamps payload")
    return TTSTimestamps(words=words, start_seconds=start_seconds, end_seconds=end_seconds)

File: packages/js/src/websocket.ts
import WebSocket from "ws";
import { decodeBase64ToBytes } from "./audio.js";
import { KovaTTSConnectionError, KovaTTSProtocolError } from "./errors.js";
import type {
  AudioChunk,
  ClientWebSocketFrame,
  ContextConfig,
  ContextClosed,
  ContextStarted,
  ErrorFrame,
  FlushCompleted,
  ResponseFormat,
  ServerWebSocketFrame,
  Timestamps,
  TTSTimestamps,
} from "./types.js";

export type StartContextOptions = {
  contextId?: string | null;
  voiceId: string;
  modelId: string;
  temperature?: number | null;
  timestamps?: boolean;
  responseFormat?: ResponseFormat;
};

export function serializeStartContext(options: StartContextOptions): ClientWebSocketFrame {
  return withOptionalContext(
    {
      start_context: {
        voice_id: options.voiceId,
        model_id: options.modelId,
        ...(options.temperature !== undefined ? { temperature: options.temperature } : {}),
        ...(options.timestamps !== undefined ? { timestamps: options.timestamps } : {}),
        ...(options.responseFormat !== undefined ? { response_format: options.responseFormat } : {}),
      },
    },
    options.contextId,
  );
}

export function serializeSendText(text: string, contextId?: string | null): ClientWebSocketFrame {
  return withOptionalContext({ send_text: text }, contextId);
}

export function serializeFlush(
  contextId?: string | null,
  flushId?: string | null,
): ClientWebSocketFrame {
  return {
    flush: true,
    ...(contextId !== undefined ? { context_id: contextId } : {}),
    ...(flushId !== undefined ? { flush_id: flushId } : {}),
  };
}

export function serializeCloseContext(
  contextId?: string | null,
  flushId?: string | null,
): ClientWebSocketFrame {
  return {
    close_context: true,
    ...(contextId !== undefined ? { context_id: contextId } : {}),
    ...(flushId !== undefined ? { flush_id: flushId } : {}),
  };
}

export function parseWebSocketFrame(value: unknown): ServerWebSocketFrame {
  if (!value || typeof value !== "object") {
    throw new KovaTTSProtocolError("WebSocket frame must be an object");
  }

  const frame = value as Record<string, unknown>;
  if (typeof frame.error === "string") {
    return {
      type: "error",
      error: frame.error,
      ...(optionalString(frame.context_id) !== undefined ? { context_id: optionalString(frame.context_id) } : {}),
      ...(optionalString(frame.flush_id) !== undefined ? { flush_id: optionalString(frame.flush_id) } : {}),
      ...(optionalString(frame.chunk_id) !== undefined ? { chunk_id: optionalString(frame.chunk_id) } : {}),
    };
  }
  if (isContextConfig(frame.context_started)) {
    return {
      type: "context_started",
      context_started: frame.context_started,
      ...(optionalString(frame.context_id) !== undefined ? { context_id: optionalString(frame.context_id) } : {}),
    };
  }
  if (typeof frame.audio_chunk === "string") {
    return {
      type: "audio",
      audio: decodeBase64ToBytes(frame.audio_chunk),
      ...(optionalString(frame.context_id) !== undefined ? { context_id: optionalString(frame.context_id) } : {}),
      ...(optionalString(frame.chunk_id) !== undefined ? { chunk_id: optionalString(frame.chunk_id) } : {}),
    };
  }
  if (frame.timestamps && typeof frame.timestamps === "object") {
    return {
      type: "timestamps",
      timestamps: frame.timestamps as TTSTimestamps,
      ...(optionalString(frame.context_id) !== undefined ? { context_id: optionalString(frame.context_id) } : {}),
      ...(optionalString(frame.chunk_id) !== undefined ? { chunk_id: optionalString(frame.chunk_id) } : {}),
    };
  }
  if (frame.flush_completed === true && typeof frame.flush_id === "string") {
    return {
      type: "flush_completed",
      flush_completed: true,
      flush_id: frame.flush_id,
      ...(optionalString(frame.context_id) !== undefined ? { context_id: optionalString(frame.context_id) } : {}),
      ...(optionalString(frame.chunk_id) !== undefined ? { chunk_id: optionalString(frame.chunk_id) } : {}),
    };
  }
  if (frame.context_closed === true) {
    return {
      type: "context_closed",
      context_closed: true,
      ...(optionalString(frame.context_id) !== undefined ? { context_id: optionalString(frame.context_id) } : {}),
    };
  }

  throw new KovaTTSProtocolError(`Unknown WebSocket frame shape: ${JSON.stringify(value)}`);
}

export function isContextStarted(frame: ServerWebSocketFrame): frame is ContextStarted {
  return frame.type === "context_started";
}

export function isAudioChunk(frame: ServerWebSocketFrame): frame is AudioChunk {
  return frame.type === "audio";
}

export function isTimestamps(frame: ServerWebSocketFrame): frame is Timestamps {
  return frame.type === "timestamps";
}

export function isFlushCompleted(frame: ServerWebSocketFrame): frame is FlushCompleted {
  return frame.type === "flush_completed";
}

export function isContextClosed(frame: ServerWebSocketFrame): frame is ContextClosed {
  return frame.type === "context_closed";
}

export function isErrorFrame(frame: ServerWebSocketFrame): frame is ErrorFrame {
  return frame.type === "error";
}

export class KovaTTSWebSocket implements AsyncIterable<ServerWebSocketFrame> {
  private readonly queue: ServerWebSocketFrame[] = [];
  private readonly waiters: Array<{
    resolve: (result: IteratorResult<ServerWebSocketFrame>) => void;
    reject: (error: Error) => void;
  }> = [];
  private closed = false;
  private closeError: Error | null = null;

  private constructor(private readonly socket: WebSocket) {
    socket.on("message", (data) => this.handleMessage(data));
    socket.on("error", (error) => this.handleError(error));
    socket.on("close", () => this.handleClose());
  }

  static connect(url: string, apiKey: string): Promise<KovaTTSWebSocket> {
    return new Promise((resolve, reject) => {
      const socket = new WebSocket(url, { headers: { "x-api-key": apiKey } });
      const fail = (error: Error) => reject(new KovaTTSConnectionError(error.message));
      socket.once("open", () => {
        socket.off("error", fail);
        resolve(new KovaTTSWebSocket(socket));
      });
      socket.once("error", fail);
    });
  }

  async startContext(options: StartContextOptions): Promise<void> {
    this.sendFrame(serializeStartContext(options));
  }

  async sendText(contextId: string | null | undefined, text: string): Promise<void> {
    this.sendFrame(serializeSendText(text, contextId));
  }

  async flush(contextId?: string | null, flushId?: string | null): Promise<void> {
    this.sendFrame(serializeFlush(contextId, flushId));
  }

  async closeContext(contextId?: string | null, flushId?: string | null): Promise<void> {
    this.sendFrame(serializeCloseContext(contextId, flushId));
  }

  close(): void {
    this.socket.close();
  }

  [Symbol.asyncIterator](): AsyncIterator<ServerWebSocketFrame> {
    return {
      next: () => this.nextFrame(),
    };
  }

  private sendFrame(frame: ClientWebSocketFrame): void {
    if (this.socket.readyState !== WebSocket.OPEN) {
      throw new KovaTTSConnectionError("WebSocket is not open");
    }
    this.socket.send(JSON.stringify(frame));
  }

  private nextFrame(): Promise<IteratorResult<ServerWebSocketFrame>> {
    if (this.queue.length > 0) {
      return Promise.resolve({ done: false, value: this.queue.shift()! });
    }
    if (this.closeError) {
      return Promise.reject(this.closeError);
    }
    if (this.closed) {
      return Promise.resolve({ done: true, value: undefined });
    }
    return new Promise((resolve, reject) => this.waiters.push({ resolve, reject }));
  }

  private handleMessage(data: WebSocket.RawData): void {
    try {
      const raw = typeof data === "string" ? data : data.toString("utf8");
      const frame = parseWebSocketFrame(JSON.parse(raw));
      if (frame.type === "error") {
        throw new KovaTTSProtocolError(frame.error);
      }
      const waiter = this.waiters.shift();
      if (waiter) {
        waiter.resolve({ done: false, value: frame });
      } else {
        this.queue.push(frame);
      }
    } catch (error) {
      this.handleError(error instanceof Error ? error : new Error(String(error)));
    }
  }

  private handleError(error: Error): void {
    this.closeError = error instanceof KovaTTSProtocolError
      ? error
      : new KovaTTSConnectionError(error.message);
    while (this.waiters.length > 0) {
      this.waiters.shift()!.reject(this.closeError);
    }
  }

  private handleClose(): void {
    this.closed = true;
    while (this.waiters.length > 0 && !this.closeError) {
      this.waiters.shift()!.resolve({ done: true, value: undefined });
    }
  }
}

function withOptionalContext<T extends object>(frame: T, contextId?: string | null): T & {
  context_id?: string | null;
} {
  return {
    ...frame,
    ...(contextId !== undefined ? { context_id: contextId } : {}),
  };
}

function isContextConfig(value: unknown): value is ContextConfig {
  return (
    !!value &&
    typeof value === "object" &&
    typeof (value as ContextConfig).voice_id === "string" &&
    typeof (value as ContextConfig).model_id === "string"
  );
}

function optionalString(value: unknown): string | null | undefined {
  if (typeof value === "string" || value === null) {
    return value;
  }
  return undefined;
}
